README.md
# A Laravel SPA SEO enhancer

PLEASE NOTE: This package is still under development.

## Example

##### Controller:

```
<?php

namespace App\Http\Controllers;

use Illuminate\Http\Request;
use Abzudev\LaravelSpaSeo\SpaContent;

class AppController extends Controller
{
    /**
     * @param Illuminate\Http\Request $request
     * @return \Illuminate\Contracts\View\Factory|\Illuminate\View\View
     */
    public function index(Request $request)
    {
        $context['html_source'] = SpaContent::getHtmlContent($request);

        return view('index', $context);
    }
}
```

##### Blade:

```
<!DOCTYPE html>
<html lang="en">
<head>
    \\ Head content
</head>
<body>

<div id="app">
    @if(isset($html_source) and !empty($html_source))
        {!! $html_source !!}
    @else
        \\ Content
    @endif
</div>

// Any additional HTML
</body>
</html>
```

##### Notes:
The package returns the rendered html of the **first** div found within the body tag. 
So to ensure that the correct html is returned, ensure the first div of the body is the mounting 
element of your SPA. After the first div you can place any additional content, but
this will not be included in the rendered html.

## Requirements

This package requires node 8.0 or higher and the Puppeteer Node library and makes use of 
[nesk/puphpeteer](https://github.com/nesk/puphpeteer) package.

You will need to install [@nesk/puphpeteer](https://www.npmjs.com/package/@nesk/puphpeteer) in your project via NPM:

```bash
npm install @nesk/puphpeteer
```

## Installation

To install this package simply run:

``` bash
composer require abzudev/laravel-spa-seo
```

If you are using Laravel 5.5+ then the package will register itself automatically.

Alternatively, you will need to add the SpaSeoServiceProvider to the list of providers in app.php.

You can publish the config-file with:

```bash
php artisan vendor:publish --provider="Abzudev\LaravelSpaSeo\SpaSeoServiceProvider" --tag="config"
```

This is the contents of the published config file:

```php
return [

    /*
    |--------------------------------------------------------------------------
    | Cache Lifetime
    |--------------------------------------------------------------------------
    |
    | Here you may specify the number of minutes that you wish the cache
    | to be allowed to remain idle before it expires.
    |
    */

    'cache_lifetime' => env('SPA_CACHE_LIFETIME', 60),

    /*
    |--------------------------------------------------------------------------
    | Route Lifetime
    |--------------------------------------------------------------------------
    |
    | This array of routes will allow you to specify the cache lifetime
    | of specific routes. By default each route will use the 'cache_lifetime'
    | as the amount of minutes that vue will be cached.
    |
    */

    'routes' => [

//        '/' => 60, (1 hour)
//        '/about' => 1440, (1 day)
//        '/contact' => 43200, (30 days)

    ],

];
```

## Testing

Run the tests with:

``` bash
composer test
```

### Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security

If you discover any security-related issues, please email jeandrew.swart@gmail.com instead of using the issue tracker.

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `laravel-spa-seo` will be documented in this file.


tests/BaseTestCase.php
<?php

namespace Abzudev\LaravelSpaSeo\Tests;

use Orchestra\Testbench\TestCase;
use Illuminate\Foundation\Application as ApplicationAlias;

class BaseTestCase extends TestCase
{
    public function setUp(): void
    {
        parent::setUp();
    }

    protected function getPackageProviders($app)
    {
        return [
            'Abzudev\LaravelSpaSeo\SpaSeoServiceProvider',
        ];
    }

    /**
     * @param ApplicationAlias $app
     */
    protected function getEnvironmentSetUp($app)
    {
        $app['config']->set('database.default', 'sqlite');
        $app['config']->set('database.connections.sqlite', [
            'driver' => 'sqlite',
            'database' => ':memory:',
            'prefix' => '',
        ]);
    }
}


tests/Unit/CacheTest.php
<?php

namespace Abzudev\LaravelSpaSeo\Tests\Unit;

use Abzudev\LaravelSpaSeo\Tests\BaseTestCase;

class CacheTest extends BaseTestCase
{
    public function setUp(): void
    {
        parent::setUp();
    }

    /** @test */
    public function a_GET_request_clears_cache()
    {
         $this->assertTrue(true);
    }

}


src/SpaContent.php
<?php

namespace Abzudev\LaravelSpaSeo;

use Carbon\Carbon;
use Illuminate\Http\Request;
use Nesk\Puphpeteer\Puppeteer;
use Sunra\PhpSimple\HtmlDomParser;

class SpaContent
{
    /**
     * @param Illuminate\Http\Request $request
     * @return array|null
     */
    public static function getHtmlContent(Request $request)
    {
        if (!$request->has('render_vue_content')) {
            return self::checkCache($request);
        }

        return null;
    }

    /**
     * @param Illuminate\Http\Request $request
     * @return mixed
     */
    public static function checkCache(Request $request)
    {
        $url = $request->fullUrlWithQuery(['render_vue_content' => 'true']);

        $route = DIRECTORY_SEPARATOR . rtrim($request->path(), DIRECTORY_SEPARATOR);

        if (array_key_exists($route, config('spaseo.routes', []))) {
            $cacheTime = config('spaseo.routes')[$route];
        } else {
            $cacheTime = config('spaseo.cache_lifetime', 60);
        }

        $cacheKey = $url . '-rendered_html';

        $content = \Cache::get($cacheKey);

        if (is_null($content)) {

            $content = self::captureHtml($url);

            \Cache::put($cacheKey, $content, Carbon::now()->addMinutes($cacheTime));
        };

        return $content;
    }

    public static function captureHtml($url)
    {
        $puppeteer = new Puppeteer();

        $browser = $puppeteer->launch();
        $page = $browser->newPage();
        $page->goto($url);
        $html = $page->content();

        $browser->close();

        $dom = HtmlDomParser::str_get_html($html);
        $domContent = $dom->find('div')[0];
        $content = $domContent->innertext();

        return $content;
    }
}


config/spaseo.php
<?php

return [

    /*
    |--------------------------------------------------------------------------
    | Cache Lifetime
    |--------------------------------------------------------------------------
    |
    | Here you may specify the number of minutes that you wish the cache
    | to be allowed to remain idle before it expires.
    |
    */

    'cache_lifetime' => env('SPA_CACHE_LIFETIME', 60),

    /*
    |--------------------------------------------------------------------------
    | Route Lifetime
    |--------------------------------------------------------------------------
    |
    | This array of routes will allow you to specify the cache lifetime
    | of specific routes. By default each route will use the 'cache_lifetime'
    | as the amount of minutes that vue will be cached.
    |
    */

    'routes' => [

//        '/' => 60, (1 hour)
//        '/about' => 1440, (1 day)
//        '/contact' => 43200, (30 days)

    ],

];
